### ISSUE TITLE
Failed requests get retried forever

### ISSUE THREAD

**Issue by jesseschalken** [bugsnag/bugsnag-agent#5]
We were recently investigating a very high traffic bill from our datacenter to Bugsnag. It turns out that all of our servers have the following in their `/var/log/upstart/bugsnag.log`:

```
...
Cannot send request. Retrying in 5 seconds
Traceback (most recent call last):
  File "/usr/bin/bugsnag-agent", line 138, in _client
    urllib2.urlopen(req).read()
  File "/usr/lib/python2.7/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 406, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 519, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 444, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 378, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 527, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 400: Bad Request
continuing...
Enqueued 590886 bytes (119/1000)
Sending 612099 bytes (118/1000)
Cannot send request. Retrying in 5 seconds
Traceback (most recent call last):
  File "/usr/bin/bugsnag-agent", line 138, in _client
    urllib2.urlopen(req).read()
  File "/usr/lib/python2.7/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 406, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 519, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 444, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 378, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 527, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 400: Bad Request
continuing...
Enqueued 573080 bytes (119/1000)
...
```

It appears to have the same set of around 120 errors stuck in it's queue, and every one of the 10 threads sends one and gets the same "400 Bad Request" response, and requeues it and waits 5 seconds. That's 2 requests per second, or 172,800 per day, or at around 1MB per request, 172GB per server per day. With around 10 servers and running `bugsnag-agent` for around 10 days, the ~10TB bandwidth bill certainly caught us off guard.
1. Is there any way to see what problem `notify.bugsnag.com` has with these requests? (they come from `bugsnag-php`)
2. Can `bugsnag-agent` be made not to retry failed requests, or limit the number of retries made for each payload?

**Comment by snmaynard** [comment]
@jesseschalken Apologies for the delay on getting back to you on this. If the requests are 1MB that seems likely that is the cause - we limit the size of a crash report and this would definitely be over that limit. Can you take a look at any of the reports to see what they contain? Do you know why they would be so large?

I think the solution of not retrying failed requests more than a few times is totally reasonable. The agent should also log the payloads that fail so we can inspect them to see what the issue is. We are totally open to a PR if you have the bandwidth to get around to this before I do - but I will try and get around to making this change this week.

**Comment by jesseschalken** [comment]
@snmaynard I was just about to deploy https://github.com/jesseschalken/bugsnag-agent.js which is just a Node.js version of the same script that doesn't retry failed requests (and doesn't have the arbitrary queue size/thread count), to at least fix our traffic problem.

I've only focused on making the agent we use not behave badly in the case of failure, and haven't bothered trying to figure out what the actual problem is with the requests.

Most payloads are in the 500KB-1MB range. What is the limit `notify.bugsnag.com` imposes?

I'll try to reproduce a failing request and get back to you. If I can't reproduce it locally I'll deploy `bugsnag-agent.js` with it logging failed payloads and have a look at the logs.

**Comment by jesseschalken** [comment]
I think the request size was indeed the problem. The threshold where `notify.bugsnag.com` responds with `401 Bad Request` is somewhere between _524,259_ and _524,534_ bytes, and according to the log the requests that are failing are consistently over 524KB and the ones the succeeding are consistently less. I found the threshold by sending a report with 1,900 stack frames through, so I still don't know why the reports happening in production are so large. It could be a big chunk of data in `$_SESSION` or something.

**Comment by jesseschalken** [comment]
I've been able to reproduce a 4MB payload. `Bugsnag_Client` has been configured to use batch sending (`$client->setBatchSending(true)`), which will queue up all errors and send them in a single request. If a lot of errors happen in a single PHP process, then the 524KB limit is easily exceeded.

Maybe `bugsnag-php` needs to be changed to bundle multiple queued errors in batches of less than 5KB?

**Comment by jesseschalken** [comment]
The large requests in production turned out to be megabytes of data stored in `$_SESSION`. See https://github.com/bugsnag/bugsnag-php/pull/91.

**Comment by kattrali** [comment]
Fixed via #14
